File: src/main/java/isel/ps/employbox/model/output/RootPage.java
package isel.ps.employbox.model.output;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import isel.ps.employbox.controllers.account.AccountController;
import isel.ps.employbox.controllers.CompanyController;
import isel.ps.employbox.controllers.jobs.JobController;
import isel.ps.employbox.controllers.UserAccountControllers.UserAccountController;

import static isel.ps.employbox.model.output.OutputDto.HOSTNAME;
import static org.springframework.hateoas.mvc.ControllerLinkBuilder.linkTo;

public class RootPage{
    @JsonProperty
    private final Self self;
    @JsonProperty
    private final AccountsCollectionLink accounts;
    @JsonProperty
    private final CompanyCollectionLink companies;
    @JsonProperty
    private final UsersCollectionLinks users;
    @JsonProperty
    private final JobsCollectionLinks jobs;
    @JsonProperty
    private final AccountLoginLink login;

    @JsonCreator
    public RootPage() {
        this.self = new Self();
        this.accounts = new AccountsCollectionLink();
        this.companies = new CompanyCollectionLink();
        this.users = new UsersCollectionLinks();
        this.jobs = new JobsCollectionLinks();
        this.login = new AccountLoginLink();
    }

    class Self{
        @JsonProperty
        final String href = HOSTNAME;
    }

    class AccountLoginLink {
        @JsonProperty
        private final _Links _links = new _Links();

        private class _Links {
            @JsonProperty
            private Self self = new Self();

            private class Self {
                @JsonProperty
                final String href = HOSTNAME + linkTo(AccountController.class).slash("self").withSelfRel().getHref();
            }
        }
    }

    class AccountsCollectionLink {
        @JsonProperty
        private final _Links _links = new _Links();

        private class _Links {
            @JsonProperty
            private Self self = new Self();

            private class Self {
                @JsonProperty
                final String href = HOSTNAME + linkTo(AccountController.class).withSelfRel().getHref();
            }
        }
    }

    class CompanyCollectionLink {
        @JsonProperty
        private final _Links _links = new _Links();

        private class _Links {
            @JsonProperty
            private Self self = new Self();

            private class Self {
                @JsonProperty
                final String href = HOSTNAME + linkTo(CompanyController.class).withSelfRel().getHref();
            }
        }
    }

    class UsersCollectionLinks {
        @JsonProperty
        private final _Links _links = new _Links();

        class _Links {
            @JsonProperty
            private Self self = new Self();

            private class Self {
                @JsonProperty
                final String href = HOSTNAME + linkTo(UserAccountController.class).withSelfRel().getHref();
            }
        }
    }

    class JobsCollectionLinks {
        @JsonProperty
        private final _Links _links = new _Links();

        class _Links {
            @JsonProperty
            private Self self = new Self();

            private class Self {
                @JsonProperty
                final String href = HOSTNAME + linkTo(JobController.class).withSelfRel().getHref();
            }
        }
    }
}

File: web-api/src/main/kotlin/isel/ps/employbox/RootPage.kt
package isel.ps.employbox

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import isel.ps.employbox.controllers.CompanyController
import isel.ps.employbox.controllers.account.AccountController
import isel.ps.employbox.controllers.jobs.JobController
import isel.ps.employbox.controllers.user.UserAccountController
import org.springframework.hateoas.mvc.ControllerLinkBuilder
import org.springframework.hateoas.mvc.ControllerLinkBuilder.linkTo


/**
 * @author tiago.ribeiro
 */
class RootPage @JsonCreator
constructor() {
    @JsonProperty
    private val self: Self
    @JsonProperty
    private val accounts: AccountsCollectionLink
    @JsonProperty
    private val companies: CompanyCollectionLink
    @JsonProperty
    private val users: UsersCollectionLinks
    @JsonProperty
    private val jobs: JobsCollectionLinks
    @JsonProperty
    private val login: AccountLoginLink

    val HOSTNAME = "http://localhost:8080"


    init {
        this.self = Self()
        this.accounts = AccountsCollectionLink()
        this.companies = CompanyCollectionLink()
        this.users = UsersCollectionLinks()
        this.jobs = JobsCollectionLinks()
        this.login = AccountLoginLink()
    }

    internal inner class Self {
        @JsonProperty
        val href = HOSTNAME
    }

    internal inner class AccountLoginLink {
        @JsonProperty
        private val _links = _Links()

        private inner class _Links {
            @JsonProperty
            private val self = Self()

            private inner class Self {
                @JsonProperty
                internal val href = HOSTNAME + ControllerLinkBuilder.linkTo(AccountController::class.java).slash("self").withSelfRel().getHref()
            }
        }
    }

    internal inner class AccountsCollectionLink {
        @JsonProperty
        private val _links = _Links()

        private inner class _Links {
            @JsonProperty
            private val self = Self()

            private inner class Self {
                @JsonProperty
                internal val href = HOSTNAME + linkTo(AccountController::class.java).withSelfRel().getHref()
            }
        }
    }

    internal inner class CompanyCollectionLink {
        @JsonProperty
        private val _links = _Links()

        private inner class _Links {
            @JsonProperty
            private val self = Self()

            private inner class Self {
                @JsonProperty
                internal val href = HOSTNAME + linkTo(CompanyController::class.java).withSelfRel().getHref()
            }
        }
    }

    internal inner class UsersCollectionLinks {
        @JsonProperty
        private val _links = _Links()

        internal inner class _Links {
            @JsonProperty
            private val self = Self()

            private inner class Self {
                @JsonProperty
                internal val href = HOSTNAME + linkTo(UserAccountController::class.java).withSelfRel().getHref()
            }
        }
    }

    internal inner class JobsCollectionLinks {
        @JsonProperty
        private val _links = _Links()

        internal inner class _Links {
            @JsonProperty
            private val self = Self()

            private inner class Self {
                @JsonProperty
                internal val href = HOSTNAME + linkTo(JobController::class.java).withSelfRel().getHref()
            }
        }
    }
}
